web3.js/src/secp256k1-program.js

```diff
@@ -0,0 +1,210 @@
+// @flow
+
+import * as BufferLayout from 'buffer-layout';
+import secp256k1 from 'secp256k1';
+import createKeccakHash from 'keccak';
+import assert from 'assert';
+
+import {PublicKey} from './publickey';
+import {TransactionInstruction} from './transaction';
+import {toBuffer} from './util/to-buffer';
+
+const {publicKeyCreate, ecdsaSign} = secp256k1;
+
+const PRIVATE_KEY_BYTES = 32;
+const PUBLIC_KEY_BYTES = 65;
+const HASHED_PUBKEY_SERIALIZED_SIZE = 20;
+const SIGNATURE_OFFSETS_SERIALIZED_SIZE = 11;
+
+/**
+ * Create a secp256k1 instruction using a public key params
+ * @typedef {Object} CreateSecp256k1InstructionWithPublicKeyParams
+ * @property {Buffer | Uint8Array | Array<number>} publicKey
+ * @property {Buffer | Uint8Array | Array<number>} message
+ * @property {Buffer | Uint8Array | Array<number>} signature
+ * @property {number} recoveryId
+ */
+export type CreateSecp256k1InstructionWithPublicKeyParams = {|
+  publicKey: Buffer | Uint8Array | Array<number>,
+  message: Buffer | Uint8Array | Array<number>,
+  signature: Buffer | Uint8Array | Array<number>,
+  recoveryId: number,
+|};
+
+/**
+ * Create a secp256k1 instruction using a private key params
+ * @typedef {Object} CreateSecp256k1InstructionWithPrivateKeyParams
+ * @property {Buffer | Uint8Array | Array<number>} privateKey
+ * @property {Buffer | Uint8Array | Array<number>} message
+ */
+export type CreateSecp256k1InstructionWithPrivateKeyParams = {|
+  privateKey: Buffer | Uint8Array | Array<number>,
+  message: Buffer | Uint8Array | Array<number>,
+|};
+
+/**
+ * A decoded Secp256k instruction
+ * @typedef {Object} DecodedSecp256k1Instruction
+ * @property {Buffer} signature
+ * @property {Buffer} ethPublicKey
+ * @property {Buffer} message
+ * @property {number} recoveryId
+ */
+export type DecodedSecp256k1Instruction = {|
+  numSignatures: number,
+  signatureOffset: number,
+  signatureInstructionOffset: number,
+  ethAddressOffset: number,
+  ethAddressInstructionIndex: number,
+  messageDataOffset: number,
+  messageDataSize: number,
+  messageInstructionIndex: number,
+  signature: Buffer,
+  ethPublicKey: Buffer,
+  recoveryId: number,
+  message: Buffer,
+|};
+
+const SECP256K1_INSTRUCTION_LAYOUT = BufferLayout.struct([
+  BufferLayout.u8('numSignatures'),
+  BufferLayout.u16('signatureOffset'),
+  BufferLayout.u8('signatureInstructionOffset'),
+  BufferLayout.u16('ethAddressOffset'),
+  BufferLayout.u8('ethAddressInstructionIndex'),
+  BufferLayout.u16('messageDataOffset'),
+  BufferLayout.u16('messageDataSize'),
+  BufferLayout.u8('messageInstructionIndex'),
+  BufferLayout.blob(20, 'ethPublicKey'),
+  BufferLayout.blob(64, 'signature'),
+  BufferLayout.u8('recoveryId'),
+]);
+
+export class Secp256k1Instruction {
+  /**
+   * Decode a secp256k1 instruction
```

Comment:
Is the intention to only decode a subset of valid secp256k1 instructions?

Reply:
That is the intention. I should get some handling for invalid instructions.

Reply:
I noticed the other programs had decoders, so I was following suit with this. Maybe it's not necessary until we need it?

Reply:
If it's not needed right now, feel free to skip. Since you already made some progress, could be nice to wrap it up. Totally up to you!

> That is the intention. Do you think I need some handling for invalid cases?

Yes, I think the decoding should be comprehensive. But the thing I was getting at is that you make a lot of assumptions about structure of the instruction which would fail to properly decode valid instructions outside of those assumptions. For example, the instruction could include multiple signatures or the message and signature order could be flipped (should be using offsets to decode these).

Reply:
Very good point about multiple signatures! The implementation I ported over to web3 of course only deals with a single signature, but it's true we need to support >1 signatures as well. I will shelve this for now.